lib/pbench/agent/tool_meister.py

```diff
@@ -200,210 +347,141 @@ def stop(self):
-                        "%s tool %s has not stopped after %d seconds",
-                        self._tool_type,
-                        self.name,
-                        5 * (6 - count),
-                    )
-        return sts
+        self.stop_process = self._create_process_with_logger(
+            args, self.tool_dir, "stop"
+        )
 
     def wait(self):
         """Wait for any tool processes to terminate after a "stop" process has
         completed.
 
         Waits for the tool's "stop" process to complete, if started, then
-        waits for the tool's start process to complete.
+        waits for the tool's start process to complete (since the "stop"
+        process is supposed stop the "start" process).
         """
+        assert self.tool_dir is not None, "Logic bomb!  no tool directory provided!"
+        if not self.tool_dir.is_dir():
+            raise RuntimeError(f"tool directory does not exist: {self.tool_dir}")
         if self.stop_process is None:
             raise ToolException(f"Tool({self.name}) wait not called after 'stop'")
         if self.start_process is None:
             raise ToolException(
                 f"Tool({self.name}) does not have a start process running"
             )
 
-        self.logger.info("Waiting for transient tool %s stop process", self.name)
-        # We wait for the stop process to finish first ...
-        sts = self._wait_for_process(self.stop_process)
-        if sts is None:
-            # The stop process did not terminate gracefully after 30 seconds,
-            # so we bring out the big guns ...
-            self.stop_process.kill()
-            self.logger.error(
-                "Killed un-responsive transient tool %s stop process after"
-                " 30 seconds",
-                self.name,
-            )
-        elif sts != 0 and sts != -(signal.SIGTERM):
-            self.logger.warning(
-                "Transient tool %s stop process failed with %d", self.name, sts
-            )
-        self.stop_process = None
-
-        # ... then we wait for the start process to finish, but we only wait
-        # for 30 seconds before we kill it.
-        self.logger.info("Waiting for transient tool %s start process", self.name)
-        try:
-            sts = self.start_process.wait(timeout=30)
-        except subprocess.TimeoutExpired:
-            self.start_process.kill()
-            self.logger.warning(
-                "Killed un-responsive transient tool %s start process after"
-                " 30 seconds",
-                self.name,
-            )
-        else:
-            if sts != 0 and sts != -(signal.SIGTERM):
-                self.logger.warning(
-                    "Transient tool %s start process failed with %d", self.name, sts
-                )
-        self.start_process = None
+        # First wait for the "stop" process to do it's job ...
+        self.stop_process = self._wait_for_process_with_kill(self.stop_process, "stop")
```

Comment:
This seems like a bad idea to me:  `_wait_for_process_with_kill()` doesn't contain an explicit `return` statement, so we're counting Python's default/implicit behavior to generate the `None` value that we want to assign to `stop_process`.  I think it would be better to treat `_wait_for_process_with_kill()` as a subroutine rather than a function, and have a separate, explicit assignment statement for `stop_process`:
```
        self._wait_for_process_with_kill(self.stop_process, "stop")
        self.stop_process = None
```
This relieves the reader from having to consult `self._wait_for_process_with_kill()` to see what value(s) `stop_process` might end up with.

Ditto for lines 376-378.

Reply:
The dangers of reviewing multiple commits.

Reply:
> The dangers of reviewing multiple commits.

Does this mean you're inclined to fix this, or should I hold my nose, or is this also fixed in #2697?

Reply:
Sorry, yes, I'll fix this.